### PR TITLE
Consistent approach to coercing and formatting date strings

### PR DESCRIPTION
**Pull request by cpsievert**
See #2402

### CODE REVIEW

File: R/utils.R [hunk 1]
```diff
@@ -1560,6 +1560,25 @@ URLencode <- function(value, reserved = FALSE) {
...
+# Also, `as.Date()` is used to coerce strings to date objects
+# so that strings like "2016-08-9" are expanded to "2016-08-09"
+dateYMD <- function(date = NULL, argName = "value") {
+  if (!length(date)) return(NULL)
+  if (length(date) > 1) warning("Expected date to be of length 1.")
+  tryCatch(format(as.Date(date), "%Y-%m-%d"),
+    error = function(e) {
+      warning(
+        "Couldn't coerce the `", argName,
+        "` argument to a date string with format yyy-mm-dd",
```
Comment (alandipert) [inline R/utils.R]: `yyy` -> `yyyy`

**Comment by wch** [comment]
Can you add a NEWS item for this?

**Comment by wch** [comment]
Two other small things:

* A test app in shiny-examples
* Please put "Closes #2402" or "Fixes #2402" in the initial comment, so that GitHub will automatically class that issue when this PR is merged.

File: NEWS.md [hunk 2]
```diff
@@ -1,6 +1,10 @@
 shiny 1.3.2.9000
 =======
 
+### Improvements
+
+* An informative warning is now thrown for mis-specified (date) strings in `dateInput()`, `updateDateInput()`, `dateRangeInput()`, and `updateDateRangeInput()`. ([#2402](https://github.com/rstudio/shiny/pull/2402))
```
Comment (wch) [inline NEWS.md]: Should start with `Fixed #2402` or "Closed" or "Resolved". Also, this PR is 2403.

**Comment by cpsievert** [comment]
@wch test app is here https://github.com/rstudio/shiny-examples/pull/139

**Pull request merged**